Implement SHA-256 in token based Remember-Me services

marcusdacoregio opened the pull request:
Closes gh-8549

<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->

<!--
Before creating new features, we recommend creating an issue to discuss the feature. This ensures that everyone is on the same page before extensive work is done.

Thanks for contributing to Spring Security. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with gh-).
-->

jzheaux commented on config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java:
```diff
@@ -186,6 +189,24 @@ public RememberMeConfigurer<H> key(String key) {
+	 * @return the {@link RememberMeConfigurer} for further customization
+	 * @since 5.7
+	 */
+	public RememberMeConfigurer<H> hashingAlgorithm(RememberMeHashingAlgorithm hashingAlgorithm) {
```
Since `RememberMeServices` can be constructed with relative ease, I'd recommend leaving this out of the DSL for the time being.

jzheaux commented on web/src/main/java/org/springframework/security/web/authentication/rememberme/TokenBasedRememberMeServices.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2004, 2005, 2006 Acegi Technology Pty Limited
```
The ones that say "Acegi Technology Pty Limited", we leave as-is

jzheaux commented on web/src/main/java/org/springframework/security/web/authentication/rememberme/TokenBasedRememberMeServices.java:
```diff
@@ -51,13 +51,20 @@
+ * or, if a hashing algorithm is configured, the form:
+ *
+ * <pre>
+ * username + &quot;:&quot; + expiryTime + &quot;:&quot; + algorithmName + &quot;:&quot;
```
I wonder if this should be thought of in the same way as how Spring Security deals with passwords.

For example, you might consider having an algorithm-for-matches and an algorithm-for-encoding, both defaulting to MD5.

* When encoding, encode with the algorithm-for-encoding, adding the prefix either way (even if the app hasn't configured it directly)
* When decoding
** if there is no algorithm in the cookie, use the algorithm-for-matches
** if there is an algorithm in the cookie, use that algorithm

This will allow folks to safely upgrade to SHA-256 without losing their old MD5 hash cookies.

jzheaux commented on web/src/main/java/org/springframework/security/web/authentication/rememberme/TokenBasedRememberMeServices.java:
```diff
@@ -83,64 +95,86 @@
+	 * @param hashingAlgorithm the hashing algorithm to use in the Cookie value
+	 */
+	public TokenBasedRememberMeServices(String key, UserDetailsService userDetailsService,
+			RememberMeHashingAlgorithm hashingAlgorithm) {
```
I think that since this is an optional parameter, it would be ideal as a setter.

jzheaux commented on docs/modules/ROOT/pages/servlet/appendix/namespace/http.adoc:
```diff
@@ -1798,6 +1798,11 @@ Specifies the period in seconds for which the remember-me cookie should be valid
 
 
+[[nsa-remember-me-hashing-algorithm]]
+* **hashing-algorithm**
```
Likewise, would recommend that this be left out for now in favor of applications using `services-ref`.

jzheaux reviewed (changes requested):
Thanks, @marcusdacoregio. I've left my comments inline.

marcusdacoregio commented:
Hi @jzheaux, I'm closing this in favor of https://github.com/spring-projects/spring-security/pull/11464. Can you please review that one? I've addressed your feedback there.
